[mlir] NamedAttribute utility generator

sjw36 opened the pull request:
All attributes in MLIR are named, inherent attributes have unscoped names and arbitrary attributes should be scoped with a dialect. Current usage is ad-hoc and much of the codebase is sprinkled with constant strings used to lookup and set attributes, leading to potential bugs when names are not updated in all usages.

This PR adds a tablegen'd utility wrapper for a NamedAttribute that manages scoped/unscoped name lookup for accessing the attribute on an Operation based on inherentness.

Dinistro reviewed (commented):
This seems a helpful extension, especially for attributes that are added to LLVM dialect operations. I would suggest to add a dedicated test for this within the test dialect, to ensure that the correctness of this is checked explicitly.

sjw36 commented:
> This seems a helpful extension, especially for attributes that are added to LLVM dialect operations. I would suggest to add a dedicated test for this within the test dialect, to ensure that the correctness of this is checked explicitly.

Thanks @Dinistro , will do.

sjw36 commented:
@joker-eph does this approach work going forward with properties? I see comments on getInherentAttr about it's transient nature.

krzysz00 commented on mlir/include/mlir/IR/AttrTypeBase.td:
```diff
@@ -283,6 +283,78 @@ class AttrDef<Dialect dialect, string name, list<Trait> traits = [],
+    static bool isInherent(::mlir::NamedAttribute attr) {
+      return attr.getName() == getMnemonic();
+    }
+    static bool has(::mlir::Operation *op) {
```
The grammar of this method name feels weird. Would it make more sense as something like `.isPresent` or `.isSetOn` or some such? Asking if an attribute has an operation feels backwards

sjw36 replied:
Yes, to follow the rest of the interface it could be `hasValue`.

krzysz00 commented on mlir/include/mlir/IR/AttrTypeBase.td:
```diff
@@ -283,6 +283,78 @@ class AttrDef<Dialect dialect, string name, list<Trait> traits = [],
+      return getScopedName();
+    }
+    // Is or Has
+    static bool is(::mlir::NamedAttribute attr) {
```
Wouldn't it make some sense to generate `is[AttrName]` out into the wider namespace? But that'd require more tablegen hackery.

joker-eph replied:
Why do we have `is` instead of `==` ? Does this mimic any existing convention on MLIR objects?

krzysz00 reviewed (commented):
Minor code-grammar nitpicks, but this is a solid utility.

joker-eph commented on mlir/include/mlir/IR/AttrTypeBase.td:
```diff
@@ -283,6 +283,78 @@ class AttrDef<Dialect dialect, string name, list<Trait> traits = [],
                                  "::" # cppClassName # ">($_self)">;
 }
 
+// Define a StringAttr wrapper for the NamedAttribute `name`
```
The "NamedAttribute" part isn't clear to me.

Also this is more than a "wrapper", it seems to me this is actually registering its own full fledge attribute? 
I'm not sure I follow the definition, but I suspect this won't be uniqued with StringAttr.

sjw36 replied:
Yes, this is no different than a StringAttr, so the usual interface works, but the new `get` only takes MLIRContext*. And once it is created, there is no way to get back to the wrapper class but it is mostly static lookup methods anyway.

joker-eph commented on mlir/include/mlir/IR/AttrTypeBase.td:
```diff
@@ -283,6 +283,78 @@ class AttrDef<Dialect dialect, string name, list<Trait> traits = [],
+
+  code namedAttrFuncs = !strconcat(typedefValueAttr, getNameFunc, [{
+    // Get name based on inherentness
+    static llvm::StringLiteral getName(Operation *op = nullptr) {
```
I don't understand the intended semantics of all this right now.

joker-eph commented on mlir/include/mlir/IR/AttrTypeBase.td:
```diff
@@ -283,6 +283,78 @@ class AttrDef<Dialect dialect, string name, list<Trait> traits = [],
+      op->setAttr(getName(op), val);
+    }
+    // Remove Value from Op
+    static void removeValue(::mlir::Operation *op) {
```
The terminology is confusing: why are you using "Value" for referring to "Attributes"?

joker-eph commented:
This is an interesting problem to solve, and there are a few aspects to it:

1) In terms of performance, we'd want to use StringAttr as much as possible to avoid string hashing and string comparisons everywhere.
2) In terms of correctness, there is the question of inherent vs discardable attributes. However this is an individual property of each operation, and the way you try to "infer" this does not seem right to me: if I have a dialect `NamedAttribute` that I intend to be `mydialect.foo`, I shouldn't lookup the inherent attribute of an operation to see if `foo` is a valid key before preprinting the `mydialect.` prefix: they are two different things!

sjw36 commented:
Thanks for the feedback @joker-eph . 

In short, it is a wrapper that is does 2 things:
1. Named lookup on objects depending on inherentness and Value Type
2. Get/Set the Typed Value for consistency 

> 1. In terms of performance, we'd want to use StringAttr as much as possible to avoid string hashing and string comparisons everywhere. But I don't think you're achieving this right now?

No, that is true but most usages are StringLiteral except when copying from a NamedAttribute. It is complicated since it is context dependent. Do you have any suggestion?

> 2. In terms of correctness, there is the question of inherent vs discardable attributes. However this is an individual property of each operation, and the way you try to "infer" this does not seem right to me: if I have a dialect `NamedAttribute` that I intend to be `mydialect.foo`, I shouldn't lookup the inherent attribute of an operation to see if `foo` is a valid key before preprinting the `mydialect.` prefix: they are two different things!

Perhaps that is a bridge too far. We have scenarios where an inherent attribute on an op is copied to the enclosing func or module and should then be dialect-qualified. I could live with a DialectNamedAttrDef and an NamedAttrDef (unqualified).

joker-eph commented:
> No, that is true but most usages are StringLiteral except when copying from a NamedAttribute. It is complicated since it is context dependent. Do you have any suggestion?

What I have been doing in the past is defining these as StringAttr on the dialect instance instead.

> Perhaps that is a bridge too far. We have scenarios where an inherent attribute on an op is copied to the enclosing func or module and should then be dialect-qualified. I could live with a DialectNamedAttrDef and an NamedAttrDef (unqualified).

The whole concept of using an "AttrDef" here is still unclear to me right now to begin with.
I'm also not sure I'm making a direct connection between your usage of inherent attribute and a propagation as a discardable one.

In general with properties we should be relying more and more on Interfaces to access informations about an operation instead of relying on "name lookup": in this model the name of the inherent attribute shouldn't matter, and shouldn't even be exposed (it is an implementation detail of the storage).

sjw36 commented:
> > No, that is true but most usages are StringLiteral except when copying from a NamedAttribute. It is complicated since it is context dependent. Do you have any suggestion?
> 
> What I have been doing in the past is defining these as StringAttr on the dialect instance instead.

Can you point me to an example. I see getConsumedAttrName() in the Transform dialect, but it is just using a StringLiteral under the hood and the usage seems cumbersome having to get loaded dialect first. Is there a better example that is storing these on the dialect?

> > Perhaps that is a bridge too far. We have scenarios where an inherent attribute on an op is copied to the enclosing func or module and should then be dialect-qualified. I could live with a DialectNamedAttrDef and an NamedAttrDef (unqualified).
> 
> The whole concept of using an "AttrDef" here is still unclear to me right now to begin with. I'm also not sure I'm making a direct connection between your usage of inherent attribute and a propagation as a discardable one.

If we set aside the inherent/property (see below) case for now and just consider a DialectNamedAttr then:
* the ODS class can be used throughout the code instead of "arbitrary_name" strings
* and the ODS class can be used to Get/Set the Typed Value on arbitrary Operations

In this way it formalizes a specific `NamedAttribute` in a Dialect with a specific Typed Value. See ROCDL example usage.

> In general with properties we should be relying more and more on Interfaces to access informations about an operation instead of relying on "name lookup": in this model the name of the inherent attribute shouldn't matter, and shouldn't even be exposed (it is an implementation detail of the storage).

This makes sense and certainly more optimal. I had a few use cases for promoting inherent attributes to parent ops but we are migrating them to properties in any case.

joker-eph commented:
> Can you point me to an example. I see getConsumedAttrName() in the Transform dialect, but it is just using a StringLiteral under the hood and the usage seems cumbersome having to get loaded dialect first. Is there a better example that is storing these on the dialect?

I don't have anything at hand: I've written this myself multiple times though. Yes you need to have a dialect pointer to get these then by nature, which is an extra step but I can't really see how to avoid though.

> the ODS class can be used throughout the code instead of "arbitrary_name" strings

How is this class any different than an "arbitrary name" though? You built a string wrapper: isn't it just like a handle to a string?

> and the ODS class can be used to Get/Set the Typed Value on arbitrary Operations

I don't quite get this part? Is this just making the `setDiscardableAttr()` API typed?
Is the whole thing here trying to make it so that the string attribute name key can only be set with a specific attribute type?

I would need to build the branch to looks a bit more, but from my read you're actually defining more than a "convenient type-safe API" and instead it is a new storage for each of these "AttrDef", even though they will be initialized with a single entry ever?

sjw36 commented:
> I don't quite get this part? Is this just making the `setDiscardableAttr()` API typed? Is the whole thing here trying to make it so that the string attribute name key can only be set with a specific attribute type?

Yes, this is a wrapper around a specific Key for the NamedAttribute and static typed accessors to the Value on an Operation.
Unfortunately I don't see a way to enforce that a particular NameAttribute must have a specified Value Type.

I cleaned it up a bit to remove the inherent check and so it is now only dialect-qualified keys for discardable attributes.
Also added unit test.

Commit:
[mlir] NamedAttribute utility generator

All attributes in MLIR are named, inherent attributes have unscoped names
and discardable attributes should be scoped with a dialect. Current usage is
ad-hoc and much of the codebase is sprinkled with constant strings used to
lookup and set attributes, leading to potential bugs when names are not
updated in all usages.

This PR adds a tablegen'd utility wrapper for a NamedAttribute that manages
scoped/unscoped name lookup for consistent typed access the attribute on an
Operation.

sjw36 commented:
Removed an unused file.

joker-eph commented:
I checked locally and this matches my expectation: it defines a full-fledge new attribute.

I guess I still don't quite get why we would do that, when it seems like some very standard C++ can achieve the same thing.
(I'll try to draft something later)

sjw36 commented:
> I checked locally and this matches my expectation: it defines a full-fledge new attribute.
> 
> I guess I still don't quite get why we would do that, when it seems like some very standard C++ can achieve the same thing. (I'll try to draft something later)

Sure, it is simple code, could easily be a template type (however, embedding the name is challenging). 
I like being about to define a specific Key and typed Value attribute that is specific to my dialect in the tablegen file with a single line.

```
def ROCDL_ReqdWorkGroupSizeAttr :
    ROCDL_NamedAttr<"ReqdWorkGroupSize", "reqd_work_group_size", "::mlir::DenseI32ArrayAttr">;
```

I did consider a new tablegen type instead of overloading AttrDef, but that seemed overly burdensome, and since the AttrDef could service the `key` aspects reasonably.

Certainly welcome a better solution. Thanks.

krzysz00 commented:
I think there's an argument for defining a new record type here. To give it a provisional name, something like `DialectScopedAttr`, so that you can generate accessors and the like in `mlir-tblgen`. 

My sense here is that there're two related usecases, which may or may not need the same abstraction:
1. Dialect-specific attributes that need to live on other people's operations. For example, that there `reqd_work_group_size`, which lives on `llvm.func`. Or `gpu.container_module`, which lives on `builtin.module`. It'll be good to have those documented and to have getter/setter functions that check invariants (since we can't inject code into other people's `verify()` methods). This sort of thing could also allow you to specify properties of the operation you put it on - for example, it could be known at tablegen time that `gpu.container_module` is supposed to go on module-like operations only.
2. Dialect-wide attribute grab bags that aren't repeated as properties on every operation. For example, stuff like `llvm.align`, `llvm.invariant_load`, and so on and so forth. In principle, we could go clean all that up and give every LLVM dialect operation a long list of properties matching exactly the kind of metadata or attributes you can put on it in LLVM IR ... or you can just have a cleaner interface to these attributes by routing them through the mechanisms for case (1) above

Making this its own tablegen record type would have the advantage of allowing you to generate documentation automatically as well, for instance, but it is a lot more work.

joker-eph commented:
> since we can't inject code into other people's verify() methods

Can you clarify what you have in mind exactly? (I don't know if you're familiar with https://mlir.llvm.org/docs/DefiningDialects/#discardable-attribute-verification ; I would assume so since you give the example of `gpu.container_module`which is exactly doing that).

> Dialect-wide attribute grab bags that aren't repeated as properties on every operation. For example, stuff like llvm.align, llvm.invariant_load, and so on and so forth. In principle, we could go clean all that up and give every LLVM dialect operation a long list of properties matching exactly the kind of metadata or attributes you can put on it in LLVM IR ... 

There is an semantic difference between "discardable" and "inherent" attributes: it's not clear to me if you're mixing these up together like if an attribute can be either of these as only a "cosmetic" thing.

> Making this its own tablegen record type would have the advantage of allowing you to generate documentation automatically as well, for instance, but it is a lot more work.

"TableGen record" is not something well defined without connecting this to a backend.
I'm mostly pointing out that using AttrDef just does not seem right to me here: I haven't perceived a requirement that would justify doing this (other than it's there and you can tweak it to get something that kind of look like providing some of the feature you want).

krzysz00 commented:
I wasn't aware of that dialect attribute verification hook - I think that's relatively new.

What I was trying to say about the LLVM dialect is that there're a lot of `llvm.*` discardable attributes these days, and it'd be good to either have them defined in a less ad-hoc way or to make them inherent attributes of relevant operations.

I'm also not sure `AttrDef` is the right mechanism here, I was pointing out motivations for describing "well-known" discardable attributes in tablegen.

joker-eph commented:
It's not really new :) https://github.com/llvm/llvm-project/commit/a495f960e0cc4d23e151288cfbfefe47d5e0c244/

joker-eph commented:
> I'll try to draft something later

Couldn't get to it before the break, but here is a draft: https://github.com/llvm/llvm-project/pull/77024

sjw36 commented:
Closing in favor of #77024.
